### PR TITLE
[Sema] Diagnose internal(set) from @inlinable functions

### PR DESCRIPTION
**Pull request by harlanhaskins**
This patch mainly consolidates the functions used to check accessors vs.
other decls, and makes sure we check setter access as well as regular
decl access.

rdar://45217648
Resolves [SR-8969](https://bugs.swift.org/browse/SR-8969)

### CODE REVIEW

**Comment by harlanhaskins** [comment]
@swift-ci please test

File: lib/Sema/ResilienceDiagnostics.cpp [hunk 1]
```diff
@@ -138,17 +138,24 @@ bool TypeChecker::diagnoseInlinableDeclRef(SourceLoc loc,
   if (downgradeToWarning == DowngradeToWarning::Yes)
     diagID = diag::resilience_decl_unavailable_warn;
 
+  // For accessors, use the name of the storage instead of the implicit '_'.
+  auto accessor = dyn_cast<AccessorDecl>(D);
+  auto printedName = accessor ?
```
Comment (jrose-apple) [inline lib/Sema/ResilienceDiagnostics.cpp]: Nitpick: this isn't really a "printed" name, since it's a DeclName.

**Comment by harlanhaskins** [comment]
@swift-ci please test

File: include/swift/AST/DiagnosticsSema.def [hunk 2]
```diff
@@ -4104,27 +4104,31 @@ ERROR(usable_from_inline_attr_in_protocol,none,
...
 
 NOTE(resilience_decl_declared_here_public,
-     none, "%0 %1 is not public", (DescriptiveDeclKind, DeclName))
+     none, DECL_OR_ACCESSOR "2 %1 is not public",
+     (DescriptiveDeclKind, DeclName, bool))
 
 NOTE(resilience_decl_declared_here,
-     none, "%0 %1 is not '@usableFromInline' or public", (DescriptiveDeclKind, DeclName))
+     none, DECL_OR_ACCESSOR "2 %1 is not '@usableFromInline' or public",
+     (DescriptiveDeclKind, DeclName, bool))
```
Comment (jrose-apple) [inline include/swift/AST/DiagnosticsSema.def]: Can you make sure each of these diagnostics is tested?

**Review by jrose-apple — APPROVED** [review]

**Comment by jrose-apple** [comment]
Note that if we can't take this on the Swift 5 branch it all gets trickier w/r/t source compat / when to use warnings vs. errors…

File: include/swift/AST/DiagnosticsSema.def (hunk 2 above)
Comment (harlanhaskins) [reply include/swift/AST/DiagnosticsSema.def]: There currently aren't any tests for the `downgradeToWarning` variant of this. Should I copy `test/attr/attr_inlinable.swift` and rewrite the `expected-error`s?

Comment (jrose-apple) [reply include/swift/AST/DiagnosticsSema.def]: Yes please! cc @slavapestov just to make sure they're not hidden somewhere.

**Comment by harlanhaskins** [comment]
@swift-ci please test

File: lib/AST/Decl.cpp [hunk 3]
```diff
@@ -2439,8 +2439,7 @@ bool ValueDecl::isUsableFromInline() const {
       return true;
 
   if (auto *containingProto = dyn_cast<ProtocolDecl>(getDeclContext())) {
-    if (isProtocolRequirement() &&
```
Comment (slavapestov) [inline lib/AST/Decl.cpp]: This change should be a separate commit with its own test case:

```
@usableFromInline protocol P {
  typealias T = Int
}

extension P {
  @inlinable func f() {
    _ = T.self
  }
}
```

**Review by slavapestov — APPROVED** [review]
Looks good except for one nit

File: include/swift/AST/DiagnosticsSema.def (hunk 2 above)
Comment (slavapestov) [reply include/swift/AST/DiagnosticsSema.def]: Take a look at `test/Compatibility/attr_inlinable*.swift`

File: test/attr/attr_inlinable.swift [hunk 4]
```diff
@@ -248,3 +248,11 @@ public struct KeypathStruct {
     // expected-error@-1 {{property 'x' is internal and cannot be referenced from an '@inlinable' function}}
   }
 }
+
+public struct HasInternalSetProperty {
+  public internal(set) var x: Int // expected-note {{setter for 'x' is not '@usableFromInline' or public}}
+
+  @inlinable public mutating func setsX() {
+    x = 10 // expected-error {{setter for 'x' is internal and cannot be referenced from an '@inlinable' function}}
```
Comment (jrose-apple) [inline test/attr/attr_inlinable.swift]: Oops, wait, we can't make this an error in 4.2 mode.

**Review by jrose-apple — CHANGES_REQUESTED** [review]

File: test/attr/attr_inlinable.swift (hunk 4 above)
Comment (jrose-apple) [reply test/attr/attr_inlinable.swift]: @slavapestov Any preference? I think we've been leaning towards always putting the latest behavior in the "normal" file and older behavior in "Compatibility".

**Review by jrose-apple — APPROVED** [review]

File: test/attr/attr_inlinable.swift (hunk 4 above)
Comment (harlanhaskins) [reply test/attr/attr_inlinable.swift]: Went ahead and just did that

**Comment by harlanhaskins** [comment]
@swift-ci please test
